tests/trainer/test_tp.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+import pytest
+import torch
+from packaging import version
+from torch.utils.data import DataLoader
+
+from composer.trainer.trainer import Trainer
+from composer.utils import dist
+from tests.common import (
+    RandomClassificationDataset,
+    SimpleModel,
+    world_size,
+)
+
+
+@pytest.mark.gpu
+@world_size(4)
+@pytest.mark.skipif(version.parse(torch.__version__) < version.parse('2.3'), reason='requires PyTorch 2.3+')
+def test_tp_train(world_size: int):
+    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel
+
+    model = SimpleModel()
+    dataset = RandomClassificationDataset(size=8)
+    dataloader = DataLoader(dataset, batch_size=2, sampler=dist.get_sampler(dataset))
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+
+    layer_plan = {
+        'fc1': ColwiseParallel(),
+        'fc2': RowwiseParallel(),
+    }
+
+    trainer = Trainer(
+        model=model,
+        optimizers=optimizer,
+        train_dataloader=dataloader,
+        parallelism_config={
+            'tp': {
+                'layer_plan': layer_plan,
+                'tensor_parallel_degree': 2,
+            },
+            'fsdp': {},
+        },
+        max_duration='3ba',
+    )
+
+    trainer.fit()
```

Comment:
How are we testing that TP actually works as expected?

Reply:
Discussed offline, will be better tested w FSDP vs. TP test. For now, this test assumes pytorch APIs work.